// ut2/lib/internals/native.js
'use strict';

exports.objectProto = Object.prototype;
exports.objectProtoToString = exports.objectProto.toString;
exports.objectProtoHasOwnProperty = exports.objectProto.hasOwnProperty;
exports.objectProtoPropertyIsEnumerable = exports.objectProto.propertyIsEnumerable;
exports.objectGetOwnPropertySymbols = Object.getOwnPropertySymbols;
exports.objectGetPrototypeOf = Object.getPrototypeOf;
exports.objectKeys = Object.keys;
exports.functionProto = Function.prototype;
exports.functionProtoToString = exports.functionProto.toString;
var symbolExisted = typeof Symbol !== 'undefined';
exports.symbolProto = symbolExisted ? Symbol.prototype : undefined;
exports.arrayProto = Array.prototype;
exports.arrayProtoSlice = exports.arrayProto.slice;
exports.mathMin = Math.min;
exports.mathMax = Math.max;
exports.mathRandom = Math.random;
exports.mathFloor = Math.floor;
exports.mathCeil = Math.ceil;
exports.mathAbs = Math.abs;
exports.numberIsFinite = Number.isFinite;
exports.numberIsInteger = Number.isInteger;
exports.numberIsSafeInteger = Number.isSafeInteger;
exports.globalThisExisted = typeof globalThis === 'object' && globalThis;
exports.globalExisted = typeof global === 'object' && global;
exports.selfExisted = typeof self === 'object' && self;
exports.MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 9007199254740991;
exports.MIN_SAFE_INTEGER = Number.MIN_SAFE_INTEGER || -9007199254740991;
exports.MAX_ARRAY_LENGTH = 4294967295;
exports.numberTag = '[object Number]';
exports.booleanTag = '[object Boolean]';
exports.stringTag = '[object String]';
exports.dateTag = '[object Date]';
exports.regExpTag = '[object RegExp]';
exports.symbolTag = '[object Symbol]';
exports.errorTag = '[object Error]';
exports.arrayBufferTag = '[object ArrayBuffer]';
exports.argumentsTag = '[object Arguments]';
exports.arrayTag = '[object Array]';
exports.typedArrayTags = ['Float32Array', 'Float64Array', 'Int8Array', 'Int16Array', 'Int32Array', 'Uint8Array', 'Uint8ClampedArray', 'Uint16Array', 'Uint32Array', 'BigInt64Array', 'BigUint64Array'].map(function (item) { return '[object ' + item + ']'; });
exports.functionTags = ['Function', 'AsyncFunction', 'GeneratorFunction', 'Proxy'].map(function (item) { return '[object ' + item + ']'; });
exports.weakSetTag = '[object WeakSet]';
exports.blobTag = '[object Blob]';
exports.domExceptionTag = '[object DOMException]';
exports.objectTag = '[object Object]';
exports.dataViewTag = '[object DataView]';
exports.mapTag = '[object Map]';
exports.promiseTag = '[object Promise]';
exports.setTag = '[object Set]';
exports.weakMapTag = '[object WeakMap]';
